ssl/quic/quic_cfq.c

```diff
@@ -0,0 +1,345 @@
+
+    if (item == NULL)
+        return NULL;
+
+    item->priority      = priority;
+    item->frame_type    = frame_type;
+    item->pn_space      = pn_space;
+    item->encoded       = (unsigned char *)encoded;
+    item->encoded_len   = encoded_len;
+    item->free_cb       = free_cb;
+    item->free_cb_arg   = free_cb_arg;
+
+    item->state = QUIC_CFQ_STATE_NEW;
+    list_remove(&cfq->free_list, item);
+    list_insert_sorted(&cfq->new_list, item, compare);
+    return &item->public;
+}
+
+void ossl_quic_cfq_mark_tx(QUIC_CFQ *cfq, QUIC_CFQ_ITEM *item)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+
+    switch (ex->state) {
+        case QUIC_CFQ_STATE_NEW:
+            list_remove(&cfq->new_list, ex);
+            list_insert_tail(&cfq->tx_list, ex);
+            ex->state = QUIC_CFQ_STATE_TX;
+            break;
+        case QUIC_CFQ_STATE_TX:
+            break; /* nothing to do */
+        default:
+            assert(0); /* invalid state (e.g. in free state) */
+            break;
+    }
+}
+
+void ossl_quic_cfq_mark_lost(QUIC_CFQ *cfq, QUIC_CFQ_ITEM *item,
+                             uint32_t priority)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+
+    switch (ex->state) {
+        case QUIC_CFQ_STATE_NEW:
+            if (priority != UINT32_MAX && priority != ex->priority) {
+                list_remove(&cfq->new_list, ex);
+                ex->priority = priority;
+                list_insert_sorted(&cfq->new_list, ex, compare);
+            }
+            break; /* nothing to do */
+        case QUIC_CFQ_STATE_TX:
+            if (priority != UINT32_MAX)
+                ex->priority = priority;
+            list_remove(&cfq->tx_list, ex);
+            list_insert_sorted(&cfq->new_list, ex, compare);
+            ex->state = QUIC_CFQ_STATE_NEW;
+            break;
+        default:
+            assert(0); /* invalid state (e.g. in free state) */
+            break;
+    }
+}
+
+/*
+ * Releases a CFQ item. The item may be in either state (NEW or TX) prior to the
+ * call. The QUIC_CFQ_ITEM pointer must not be used following this call.
+ */
+void ossl_quic_cfq_release(QUIC_CFQ *cfq, QUIC_CFQ_ITEM *item)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+    switch (ex->state) {
+        case QUIC_CFQ_STATE_NEW:
+            list_remove(&cfq->new_list, ex);
+            list_insert_tail(&cfq->free_list, ex);
+            clear_item(ex);
+            break;
+        case QUIC_CFQ_STATE_TX:
+            list_remove(&cfq->tx_list, ex);
+            list_insert_tail(&cfq->free_list, ex);
+            clear_item(ex);
+            break;
+        default:
+            assert(0); /* invalid state (e.g. in free state) */
+            break;
+    }
+}
+
+QUIC_CFQ_ITEM *ossl_quic_cfq_get_priority_head(QUIC_CFQ *cfq, uint32_t pn_space)
+{
+    QUIC_CFQ_ITEM_EX *item = cfq->new_list.head;
+
+    for (; item != NULL && item->pn_space != pn_space; item = item->next);
```

Comment:
Is this likely to iterate much?

If so, we should investigate a skip list or similar.

Reply:
Yeah, I'm not concerned about excessive usage here. The CFQ is for now only going to be used for NEW_CONN_ID, RETIRE_CONN_ID, NEW_TOKEN, barring any QUIC extensions.